WalletWasabi.Tests/UnitTests/Blockchain/TransactionOutputs/CoinsRegistryTests.cs

```diff
@@ -0,0 +1,182 @@
+		//                  |
+		//                  +--tx3 (replacement)---> (E)
+
+		Coin? tx0Coin;
+		SmartTransaction tx0;
+		SmartTransaction tx1;
+		SmartTransaction tx2;
+		SmartTransaction tx3;
+
+		// Create and process transaction tx0.
+		{
+			tx0 = CreateCreditingTransaction(NewInternalKey(label: "A").P2wpkhScript, Money.Coins(1.0m), height: 54321);
+			tx0Coin = tx0.Transaction.Outputs.AsCoins().First();
+
+			IReadOnlyList<SmartCoin> tx0Coins = ProcessTransaction(tx0);
+			_ = Assert.Single(tx0Coins);
+			_ = Assert.Single(Coins);
+			Assert.Equal(tx0Coins[0], Coins.First());
+		}
+
+		// Create and process transaction tx1 that fully spends tx0.
+		{
+			tx1 = CreateSpendingTransaction(tx0Coin, NewInternalKey(label: "B").P2wpkhScript);
+			tx1.Transaction.Inputs[0].Sequence = Sequence.OptInRBF;
+			tx1.Transaction.Outputs[0].Value = Money.Coins(0.95m);
+			tx1.Transaction.Outputs.Add(Money.Coins(0.1m), NewInternalKey("C").P2wpkhScript);
+
+			IReadOnlyList<SmartCoin> tx1Coins = ProcessTransaction(tx1);
+			AssertEqualCoinSets(Coins, tx1Coins);
+
+			SmartCoin unconfirmedCoin1 = Assert.Single(Coins, coin => coin.HdPubKey.Labels == "B");
+			SmartCoin unconfirmedCoin2 = Assert.Single(Coins, coin => coin.HdPubKey.Labels == "C");
+			Assert.True(unconfirmedCoin1.IsReplaceable());
+			Assert.True(unconfirmedCoin2.IsReplaceable());
+
+			Assert.True(Coins.IsKnown(tx0.GetHash()));
+			Assert.True(Coins.IsKnown(tx1.GetHash()));
+		}
+
+		// Create and process transaction tx2 that partially spends tx1.
+		{
+			SmartCoin coin = Assert.Single(Coins, coin => coin.HdPubKey.Labels == "B");
+
+			tx2 = CreateSpendingTransaction(coin.Coin, NewInternalKey("D").P2wpkhScript);
+			tx2.Transaction.Outputs[0].Value = Money.Coins(0.7m);
+
+			IReadOnlyList<SmartCoin> tx2Coins = ProcessTransaction(tx2);
+			_ = Assert.Single(tx2Coins);
+
+			Assert.True(Coins.IsKnown(tx1.GetHash()));
+			Assert.True(Coins.IsKnown(tx2.GetHash()));
+		}
+
+		// Create and process REPLACEMENT transaction tx3 that fully spends tx0.
+		{
+			// Undo tx1.
+			{
+				Assert.True(Coins.IsKnown(tx0.GetHash()));
+				Assert.True(Coins.IsKnown(tx1.GetHash()));
+				Assert.True(Coins.IsKnown(tx2.GetHash()));
+
+				// First undo tx1 which should transitively undo tx2.
+				(ICoinsView toRemove, ICoinsView toAdd) = Coins.Undo(tx1.GetHash());
+
+				Assert.True(Coins.IsKnown(tx0.GetHash()));
+				Assert.False(Coins.IsKnown(tx1.GetHash()));
+				// Assert.False(Coins.IsKnown(tx2.GetHash())); // Bug as of now.
+			}
+
+			// .. then create and process tx3.
+			{
+				tx3 = CreateSpendingTransaction(tx0Coin, NewInternalKey("E").P2wpkhScript);
+				tx3.Transaction.Outputs[0].Value = Money.Coins(0.9m);
+				IReadOnlyList<SmartCoin> tx3Coins = ProcessTransaction(tx3);
+
+				SmartCoin finalCoin = Assert.Single(Coins);
+				Assert.Equal("E", finalCoin.HdPubKey.Labels);
+
+				Assert.Empty(Coins.AsAllCoinsView().Where(coin => coin.HdPubKey.Labels == "B"));
```

Comment:
Nit: Maybe we could check for "C" and "D", too.
```suggestion
				Assert.Empty(Coins.AsAllCoinsView().Where(coin => coin.HdPubKey.Labels == "B"));
				Assert.Empty(Coins.AsAllCoinsView().Where(coin => coin.HdPubKey.Labels == "C"));
				Assert.Empty(Coins.AsAllCoinsView().Where(coin => coin.HdPubKey.Labels == "D"));
```